File: 0015_3sum.py
# https://leetcode-cn.com/problems/3sum/

class Solution:
    def threeSum(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: list[list[int]]
        """
        
        n = len(nums)
        if n < 3:
            return []
        
        res = []
        nums.sort()
        for i in range(n):
            if nums[i] > 0:
                break
            
            if i > 0 and nums[i] == nums[i-1]:
                continue
            
            left = i + 1
            right = n - 1
            while left < right:
                sum_3 = nums[i] + nums[left] + nums[right]
                if sum_3 == 0:
                    res.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left+1]:
                        left = left + 1
                    while left < right and nums[right] == nums[right-1]:
                        right = right - 1
                    left = left + 1
                    right = right - 1
                elif sum_3 > 0:
                    right = right - 1
                else:
                    left = left + 1
            return res

File: Cpp/0015_3sum.cpp
// https://leetcode-cn.com/problems/3sum/

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;

        int n = nums.size();
        if (n < 3)
            return res;

        std::sort(nums.begin(), nums.end());
        if ((nums[0] > 0) || (nums[n-1] < 0))
            return res;

        for (int i = 0; i < n-2; ++i) {
            int nums_i = nums[i];
            if ((i > 0) && (nums_i == nums[i-1]))
                continue;

            int j = i + 1, k = n - 1;
            while (j < k) {
                int nums_j = nums[j], nums_k = nums[k];

                // Ignore duplicate elements.
                if ((j > i+1) && (nums_j == nums[j-1])) {
                    j = j + 1;
                    continue;
                }
                if ((k < n-1) && (nums_k == nums[k+1])) {
                    k = k - 1;
                    continue;
                }

                int sum_ijk = nums_i + nums_j + nums_k;
                if (sum_ijk == 0) {
                    vector<int> tmp = {nums_i, nums_j, nums_k};
                    res.push_back(tmp);
                    j = j + 1;
                    k = k - 1;
                }
                else if (sum_ijk < 0)
                    j = j + 1;
                else
                    k = k - 1;
            }
        }

        return res;
    }
};